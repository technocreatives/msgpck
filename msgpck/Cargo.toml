[package]
name = "msgpck"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
num-traits = { version = "0.2.17", default-features = false }
msgpck_derive = { version = "0.1.0", path = "../msgpck_derive", optional = true }
heapless07 = { package = "heapless", version = "0.7.16", optional = true }
thiserror = { version = "1.0.50", optional = true }

[features]
default = ["std", "derive", "heapless07"]
heapless07 = ["dep:heapless07"]
alloc = []
debug = []
std = ["debug", "alloc", "num-traits/std", "dep:thiserror"]
derive = ["dep:msgpck_derive"]

[dev-dependencies]
paste = "1.0.14"
proptest = "1.3.1"
proptest-derive = "0.4.0"
pretty-hex = "0.3.0"
