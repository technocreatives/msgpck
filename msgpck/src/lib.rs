//! A light-weight library for serializing/deserializing types as
//! [MessagePack](https://msgpack.org/) ("msgpack").
//!
//! The goal of this library is to be as light-weight as possible, while still
//! being easy to use. This is especially useful for embedded systems, where
//! binary size is important.
//!
//! To limit binary bloat, this library doesn't use serde. Instead, we provide
//! the [MsgPck] and [UnMsgPck] traits, which can be derived for most types.
//!
//! # Usage
//!
//! The easiest way to use this library is by deriving the [MsgPck] and
//! [UnMsgPck] traits on your types. This will automatically generate the
//! necessary code to serialize and deserialize your types.
//!
//! ```rust
//! use msgpck::{MsgPck, UnMsgPck};
//!
//! # #[derive(Debug, PartialEq)]
//! #[derive(MsgPck, UnMsgPck)]
//! struct Foo {
//!     a: bool,
//!     b: String,
//! }
//!
//! let foo = Foo { a: true, b: "hello".to_owned() };
//! let bytes = msgpck::pack_vec(&foo).unwrap();
//! # let foo2: Foo = msgpck::unpack_bytes(&bytes).unwrap();
//! # assert_eq!(foo, foo2);
//! ```
//!
//! ## Custom implementation
//!
//! Here is an example of a custom implementation for a unit struct:
//!
//! ```
//! # use proptest::prelude::*;
//! # use proptest_derive::Arbitrary;
//! use msgpck::{MsgPck, UnMsgPck, MsgWriter, PackError, Marker, UnpackError, slice_take};
//!
//! # #[derive(Arbitrary, Debug, PartialEq)]
//! struct User { id: u32, name: String }
//!
//! impl MsgPck for User {
//!     fn pack(&self, writer: &mut dyn MsgWriter) -> Result<(), PackError> {
//!         // Structs are serialized as arrays of their fields
//!         writer.write(&[Marker::FixArray(2).to_u8()])?;
//!         // serialize the fields
//!         self.id.pack(writer)?;
//!         self.name.pack(writer)?;
//!         Ok(())
//!     }
//! }
//!
//! // The lifetime `'buf` is the lifetime of the buffer we're reading from
//! impl<'buf> UnMsgPck<'buf> for User {
//!     fn unpack(source: &mut &'buf [u8]) -> Result<Self, UnpackError>
//!     where
//!         Self: Sized,
//!     {
//!         // Use helper function to read the first byte
//!         let &[b] = slice_take(source).map_err(|_| dbg!(UnpackError::UnexpectedEof))?;
//!
//!         // Check that it's the marker for a 2-element array
//!         let m = Marker::from_u8(b);
//!         if !matches!(m, Marker::FixArray(2)) {
//!             return Err(UnpackError::WrongMarker(m));
//!         };
//!
//!         // Unpack the fields
//!         Ok(User {
//!             id: u32::unpack(source)?,
//!             name: String::unpack(source)?
//!         })
//!     }
//! }
//! #
//! # proptest! {
//! #     #[test]
//! #     fn roundtrip(s: User) {
//! #         let mut writer: Vec<_> = Vec::new();
//! #         s.pack(&mut writer).unwrap();
//! #         let d = User::unpack(&mut &writer[..]).unwrap();
//! #         assert_eq!(s, d);
//! #     }
//! # }
//! ```
//!
//! # Compatibility with `rmp_serde`
//!
//! We aim to be able to deserialize any value serialized using rmp_serde.
//!
//! This means we use the following patterns for serializing:
//!
//! - `struct`s are serialized as arrays of their fields
//! - An `enum` variants with no fields is serialized as a string
//! - An `enum` variant with one field is a map `{ variant_name: field_value }`
//! - An `enum` variant with multiple fields is a map `{ variant_name: [field_valuesâ€¦] }`
// TODO: - `Option<T>` is serialized as an array of 0 or 1 elements
//!
// *TODO: decide if we're gonna change serialized representation of enums*

#![cfg_attr(not(feature = "std"), no_std)]

mod impls;
mod marker;
mod pack;
mod unpack;
mod utils;

pub use crate::{
    impls::{EnumHeader, Variant},
    marker::Marker,
    pack::{
        errors::PackError,
        helpers::{pack_slice, pack_vec},
        size_hint::SizeHint,
        writers::{BufferWriter, MsgWriter, WriteError},
        MsgPck,
    },
    unpack::{errors::UnpackError, helpers::unpack_bytes, UnMsgPck},
    utils::{pack_array_header, slice_take, unpack_array_header, unpack_enum_header},
};

#[cfg(feature = "derive")]
pub use msgpck_derive::{MsgPck, UnMsgPck};
