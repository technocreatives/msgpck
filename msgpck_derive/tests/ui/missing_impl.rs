use msgpck::{MsgPck, UnMsgPck};

struct Foo {
    a: bool,
    b: String,
}

#[derive(MsgPck, UnMsgPck)]
struct Bar {
    a: f32,
    foo: Foo,
}
