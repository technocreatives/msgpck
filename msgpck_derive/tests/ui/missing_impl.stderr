error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/ui/missing_impl.rs:12:2
   |
12 | }
   |  ^ consider adding a `main` function to `$DIR/tests/ui/missing_impl.rs`

error[E0277]: the trait bound `Foo: MsgPck` is not satisfied
  --> tests/ui/missing_impl.rs:11:5
   |
11 |     foo: Foo,
   |     ^^^^^^^^ the trait `MsgPck` is not implemented for `Foo`
   |
   = help: the following other types implement trait `MsgPck`:
             bool
             isize
             i8
             i16
             i32
             i64
             usize
             u8
           and $N others

error[E0277]: the trait bound `Foo: UnMsgPck<'_>` is not satisfied
  --> tests/ui/missing_impl.rs:11:5
   |
11 |     foo: Foo,
   |     ^^^^^^^^ the trait `UnMsgPck<'_>` is not implemented for `Foo`
   |
   = help: the following other types implement trait `UnMsgPck<'buf>`:
             <bool as UnMsgPck<'buf>>
             <isize as UnMsgPck<'buf>>
             <i8 as UnMsgPck<'buf>>
             <i16 as UnMsgPck<'buf>>
             <i32 as UnMsgPck<'buf>>
             <i64 as UnMsgPck<'buf>>
             <usize as UnMsgPck<'buf>>
             <u8 as UnMsgPck<'buf>>
           and $N others
